File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/AdminBroadcastCommand.cs
using Atlasd.Daemon;
using Atlasd.Localization;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class AdminBroadcastCommand : ChatCommand
    {
        public AdminBroadcastCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            Task.Run(() =>
            {
                var maskBroadcaster = Settings.GetBoolean(new string[] { "battlenet", "emulation", "mask_admins_in_broadcasts" }, false);

                var broadcasterFlags = maskBroadcaster ? Account.Flags.Admin : context.GameState.ChannelFlags;
                var broadcasterPing = maskBroadcaster ? -1 : context.GameState.Ping;
                var broadcasterName = maskBroadcaster ? Settings.GetString(new string[] { "battlenet", "realm", "name" }, Resources.Battlenet) : context.GameState.OnlineName;

                var chatEvent = new ChatEvent(ChatEvent.EventIds.EID_BROADCAST, broadcasterFlags, broadcasterPing, broadcasterName, RawBuffer);
                foreach (var gameState in Battlenet.Common.ActiveGameStates.Values) chatEvent.WriteTo(gameState.Client);
            });
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/AdminBroadcastCommand.vb
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic
Imports System.Threading.Tasks

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class AdminBroadcastCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Task.Run(Sub()
                         Dim lamChatEvent = New ChatEvent(ChatEvent.EventIds.EID_BROADCAST, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, RawBuffer)

                         SyncLock Battlenet.Common.ActiveGameStates
                             For Each pair In Battlenet.Common.ActiveGameStates
                                 lamChatEvent.WriteTo(pair.Value.Client)
                             Next
                         End SyncLock
                     End Sub)
        End Sub
    End Class
End Namespace
